timeinterval/interval.go

```diff
@@ -0,0 +1,101 @@
+package timeinterval
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+)
+
+// Interval describes an interval bounded by a StartsAt and EndsAt time.
+// the unexported "iso8601" is used to store the user's ISO8601 string. This makes it possible to marshal/unmarshal
+// the interval to/from the same ISO8601 representation originally provided if desired.
+type Interval struct {
+	iso8601  string
+	StartsAt time.Time
+	EndsAt   time.Time
+}
+
+// NewInterval returns an Interval instance based upon the given options
+// and an error if the given options are insufficient to construct the interval.
+func NewInterval(startsAt, endsAt *time.Time, duration *time.Duration, iso8601 *string) (*Interval, error) {
```

Comment:
`func NewInterval(startsAt, endsAt *time.Time) (*Interval, error) {`

Should be fine. That will be pretty self-explanatory.

Add a `func (in Interval) Validate() error` It should check that start < end and both are not nil (which is invalid I suspect).